File: c/graph/shortest_path/Dijkstra.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef struct node {
    int index;
    int weight;
    struct node *next;
} EdgeNode;

typedef struct {
    int value;
    EdgeNode *next;
} VertexNode;

typedef VertexNode *pNode;

typedef struct {
    int *values;
    int length;
} ArrayNode;

typedef struct {
    int index;
    int weight;
} HeapNode;

typedef struct {
    HeapNode *values;
    int length;
} QueueNode;

static QueueNode *createQueue(int length);
static void enqueue(QueueNode *queue, int index, int weight);
static HeapNode dequeue(QueueNode *queue);

void dijkstra(pNode *graphs, int length, int start, ArrayNode *result) {
    int *visited = (int *) malloc(sizeof(int) * length);
    memset(visited, 0, sizeof(int) * length);
    QueueNode *queue = createQueue(length);
    enqueue(queue, start, 0);

    result->values = (int *) malloc(sizeof(int) * length);
    for (int i = 0; i < length; ++i) {
        result->values[i] = ~(1 << 31);
    }
    result->length = length;
    while (queue->length) {
        HeapNode heap = dequeue(queue);
        int index = heap.index;
        int weight = heap.weight;
        if (visited[index]) {
            continue;
        }

        result->values[index] = weight;
        visited[index] = 1;

        EdgeNode *edge = graphs[index]->next;
        while (edge) {
            int edgeIndex = edge->index;
            int edgeWeight = weight + edge->weight;
            if (!visited[edgeIndex]) {
                enqueue(queue, edgeIndex, edgeWeight);
            }
            edge = edge->next;
        }
    }
}

static QueueNode *createQueue(int length) {
    QueueNode *queue = (QueueNode *) malloc(sizeof(QueueNode));
    queue->values = (HeapNode *) malloc(sizeof(HeapNode) * length);
    queue->length = 0;
    return queue;
}

static void enqueue(QueueNode *queue, int index, int weight) {
    int child = queue->length;
    while (child > 0) {
        int parent = (child - 1) >> 1;
        if (queue->values[parent].weight <= weight) {
            break;
        }
        queue->values[child] = queue->values[parent];
        child = parent;
    }
    queue->values[child].index = index;
    queue->values[child].weight = weight;
    ++queue->length;
}

static HeapNode dequeue(QueueNode *queue) {
    --queue->length;
    HeapNode value = queue->values[0];
    HeapNode pivot = queue->values[queue->length];
    int parent = 0;
    while (1) {
        int child = (parent << 1) + 1;
        if (child >= queue->length) {
            break;
        }

        if (child < queue->length - 1 && queue->values[child].weight > queue->values[child + 1].weight) {
            ++child;
        }
        if (queue->values[child].weight >= pivot.weight) {
            break;
        }
        queue->values[parent] = queue->values[child];
        parent = child;
    }
    queue->values[parent] = pivot;
    return value;
}

File: py/graph/minimum_cost_maximum_flow/Dijkstra.py
#!/usr/bin/env python
# -*- coding:gbk -*-

import heapq, sys
from typing import List

class EdgeNode:
    def __init__(self, index: int, weight: int, price: int):
        self.index, self.weight, self.price, self.next = index, weight, price, None

class VertexNode:
    def __init__(self, value: int):
        self.value, self.next = value, None

class Dijkstra:
    def search(self, graphs: List[VertexNode], start: int, end: int) -> List[int]:
        weights = [[0 for _ in range(len(graphs))] for _ in range(len(graphs))]
        for i, value in enumerate(graphs):
            edge = value.next
            while edge:
                weights[i][edge.index] = edge.weight
                edge = edge.next

        parent, maxFlow, minCost, offset, result = [], [], [], [0 for _ in range(len(graphs))], [0 for _ in range(2)]
        while self.__search(graphs, start, end, weights, parent, maxFlow, minCost, offset):
            result[0] += maxFlow[end]
            result[1] += maxFlow[end] * (minCost[end] - offset[start] + offset[end])

            index = end
            while index != start:
                weights[parent[index]][index] -= maxFlow[end]
                weights[index][parent[index]] += maxFlow[end]
                index = parent[index]
            for i, value in enumerate(minCost):
                offset[i] += value
        return result

    def __search(self, graphs: List[VertexNode], start: int, end: int, weights: List[List[int]], parent: List[int], maxFlow: List[int], minCost: List[int], offset: List[int]) -> bool:
        parent[ : ], maxFlow[ : ], minCost[ : ] = [-1 for _ in range(len(graphs))], [sys.maxsize for _ in range(len(graphs))], [sys.maxsize for _ in range(len(graphs))]
        minCost[start] = 0

        visited, queue = [False for _ in range(len(graphs))], [self.__HeapNode(start, minCost[start])]
        while queue:
            heap = heapq.heappop(queue)
            index = heap.index
            visited[index], edge = True, graphs[index].next
            while edge:
                edgeIndex = edge.index
                if weights[index][edgeIndex] and not visited[edgeIndex] and minCost[edgeIndex] > minCost[index] + edge.price + offset[index] - offset[edgeIndex]:
                    parent[edgeIndex], maxFlow[edgeIndex], minCost[edgeIndex] = index, min(maxFlow[index], weights[index][edgeIndex]), minCost[index] + edge.price + offset[index] - offset[edgeIndex]
                    heapq.heappush(queue, self.__HeapNode(edgeIndex, minCost[edgeIndex]))
                edge = edge.next
        return sys.maxsize != minCost[end]

    class __HeapNode:
        def __init__(self, index: int, price: int):
            self.index, self.price = index, price

        def __lt__(self, other):
            return self.price < other.price
